### PR TITLE
clean up websocket stuff

### PR DESCRIPTION
**Pull request by GusCaplan**

### CODE REVIEW

File: src/client/websocket/WebSocketConnection.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  if (browser) return window.WebSocket; // eslint-disable-line no-undef
+  try {
+    return require('uws');
+  } catch (e) {
+    return require('ws');
+  }
+}());
+
+
+class WebSocketConnection extends EventEmitter {
```
Comment (Gawdl3y) [inline src/client/websocket/WebSocketConnection.js]: Missing documentation for the entire class.

File: src/client/websocket/WebSocketConnection.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+  /**
+   * Called when the websocket gets a message
+   * @param {Object} event Close event object
+   * @returns {Promise<boolean>}
+   */
+  eventMessage(event) {
+    if (this.listenerCount('message')) this.emit('message', event);
+    return this.unpack(event.data)
+    .then(data => {
```
Comment (Gawdl3y) [inline src/client/websocket/WebSocketConnection.js]: The `then` and `catch` calls should be moved to the closing line of the previous call.

File: src/client/websocket/WebSocketConnection.js [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+      });
+    });
+  }
+
+  /**
+   * The encoding this connection will use
+   * @type {string}
+   * @static
+   */
+  static get encoding() {
```
Comment (Gawdl3y) [inline src/client/websocket/WebSocketConnection.js]: This should be a regular static property called `ENCODING`, not a getter (and all caps). `@static` will be unnecessary.

File: src/client/websocket/WebSocketConnection.js [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    return erlpack ? 'etf' : 'json';
+  }
+}
+
+WebSocketConnection.CONNECTING = 0;
+WebSocketConnection.OPEN = 1;
+WebSocketConnection.CLOSING = 2;
+WebSocketConnection.CLOSED = 3;
+
```
Comment (Gawdl3y) [inline src/client/websocket/WebSocketConnection.js]: Each of these needs to be documented.

File: src/client/websocket/WebSocketConnection.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  if (browser) return window.WebSocket; // eslint-disable-line no-undef
+  try {
+    return require('uws');
+  } catch (e) {
+    return require('ws');
+  }
+}());
+
+/**
+ * Abstracts a WebSocket connection with decoding/encoding for the discord gateway
```
Comment (Gawdl3y) [inline src/client/websocket/WebSocketConnection.js]: Should be marked `@private`.

File: src/client/ClientManager.js [hunk 6]
```diff
@@ -29,7 +30,8 @@ class ClientManager {
     this.client.emit(Constants.Events.DEBUG, `Authenticated using token ${token}`);
     this.client.token = token;
     const timeout = this.client.setTimeout(() => reject(new Error(Constants.Errors.TOOK_TOO_LONG)), 1000 * 300);
-    this.client.rest.methods.getGateway().then(gateway => {
+    this.client.rest.methods.getGateway(this.client.options.shardCount === 'auto').then(res => {
```
Comment (GusCaplan) [inline src/client/ClientManager.js]: nah

**Comment by iCrawl** [comment]
The initial connect works now, but on reconnect it throws:

**Pull request closed without merging**